zebra-test/src/command.rs

```diff
@@ -847,6 +848,92 @@ impl<T> TestChild<T> {
         }
     }
 
+    /// Checks each line of the child's stdout, until it finds every regex in `success_regexes`,
+    /// and returns all the lines matched by any regex. Prints all stdout lines.
+    ///
+    /// Kills the child on error, or after the configured timeout has elapsed.
+    /// See [`Self::expect_line_matching_regex_set`] for details.
+    //
+    // TODO: these methods could block if stderr is full and stdout is waiting for stderr to be read
+    #[instrument(skip(self))]
+    #[allow(clippy::unwrap_in_result)]
+    pub fn expect_stdout_line_matches_all_unordered<RegexList>(
+        &mut self,
+        unordered_regexes: RegexList,
+    ) -> Result<Vec<String>>
+    where
+        RegexList: IntoIterator + Debug,
+        RegexList::Item: ToRegexSet,
+    {
+        let regex_list = unordered_regexes.collect_regex_set()?;
+
+        let mut unmatched_indexes: HashSet<usize> = (0..regex_list.len()).collect();
+        let mut matched_lines = Vec::new();
+
+        while !unmatched_indexes.is_empty() {
+            let line = self
+                .expect_stdout_line_matches(regex_list.clone())
+                .map_err(|err| {
+                    let unmatched_regexes = regex_list.patterns_for_indexes(&unmatched_indexes);
+
+                    err.with_section(|| {
+                        format!("{unmatched_regexes:#?}").header("Unmatched regexes:")
+                    })
+                    .with_section(|| format!("{matched_lines:#?}").header("Matched lines:"))
+                })?;
+
+            let matched_indices: HashSet<usize> = regex_list.matches(&line).iter().collect();
+            unmatched_indexes = &unmatched_indexes - &matched_indices;
+
+            matched_lines.push(line);
+        }
+
+        Ok(matched_lines)
+    }
+
+    /// Checks each line of the child's stderr, until it finds every regex in `success_regexes`,
+    /// and returns all the lines matched by any regex. Prints all stderr lines.
```

Comment:
If I understand the method correctly, it would be more accurate to say that it returns the first matching line for each regex:

```suggestion
    /// and returns the first matching line for each regex or a list of unmatched regexes.
    /// Prints all stdout lines.
```

Reply:
```suggestion
    /// and returns all lines matched by any regex, until each regex has been matched at least once.
    /// If the output finishes or the command times out before all regexes are matched, returns an error with
    /// a list of unmatched regexes. Prints all stderr lines.
```